File: tramaj-js/src/node.ts
/**
 * `Node`/`NodeAttribute` and the strict `specs/node-json.md` encode/decode,
 * hand-written rather than derived so the "a decoder MUST reject ... MUST NOT
 * infer a missing field from a default" rules are enforced exactly.
 */

import { hasField, isJsonObject, jsonObjectFrom, type Json, type JsonObject } from "./json.js";

export type Annotations = Record<string, Json>;

export function noAnnotations(): Annotations {
  return {};
}

export type Node =
  | { t: "text"; value: Json; annotations: Annotations }
  | {
      t: "element";
      tag: string;
      attributes: NodeAttribute[];
      value: Json;
      children: Node[];
      annotations: Annotations;
    }
  | { t: "fragment"; children: Node[]; annotations: Annotations };

export type NodeAttribute =
  | { t: "attribute"; name: string; value: Json }
  | { t: "action"; event: string; key: string; payload: Json };

export function textNode(value: Json, annotations: Annotations = noAnnotations()): Node {
  return { t: "text", value, annotations };
}

export function elementNode(
  tag: string,
  attributes: NodeAttribute[],
  value: Json,
  children: Node[],
  annotations: Annotations = noAnnotations(),
): Node {
  return { t: "element", tag, attributes, value, children, annotations };
}

export function fragmentNode(children: Node[], annotations: Annotations = noAnnotations()): Node {
  return { t: "fragment", children, annotations };
}

function annotationsToJson(anns: Annotations): JsonObject {
  return jsonObjectFrom(Object.keys(anns).map((k) => [k, anns[k] as Json] as const));
}

export function nodeToJson(n: Node): Json {
  switch (n.t) {
    case "text":
      return { type: "text", value: n.value, annotations: annotationsToJson(n.annotations) };
    case "element":
      return {
        type: "element",
        tag: n.tag,
        attributes: n.attributes.map(nodeAttributeToJson),
        value: n.value,
        children: n.children.map(nodeToJson),
        annotations: annotationsToJson(n.annotations),
      };
    case "fragment":
      return {
        type: "fragment",
        children: n.children.map(nodeToJson),
        annotations: annotationsToJson(n.annotations),
      };
  }
}

export function nodeAttributeToJson(a: NodeAttribute): Json {
  return a.t === "attribute"
    ? { kind: "attribute", name: a.name, value: a.value }
    : { kind: "action", event: a.event, key: a.key, payload: a.payload };
}

export class NodeDecodeError extends Error {
  override readonly name = "NodeDecodeError";
}

function req(what: string, field: string, obj: JsonObject): Json {
  if (!hasField(obj, field)) {
    throw new NodeDecodeError(`${what}: missing required field ${JSON.stringify(field)}`);
  }
  return obj[field] as Json;
}

function reqString(what: string, field: string, obj: JsonObject): string {
  const v = req(what, field, obj);
  if (typeof v !== "string") {
    throw new NodeDecodeError(`${what}: field ${JSON.stringify(field)} must be a string`);
  }
  return v;
}

function reqArray(what: string, field: string, obj: JsonObject): Json[] {
  const v = req(what, field, obj);
  if (!Array.isArray(v)) {
    throw new NodeDecodeError(`${what}: field ${JSON.stringify(field)} must be an array`);
  }
  return v;
}

function reqAnnotations(obj: JsonObject): Annotations {
  const v = req("node", "annotations", obj);
  if (!isJsonObject(v)) {
    throw new NodeDecodeError('node: field "annotations" must be an object');
  }
  return jsonObjectFrom(Object.keys(v).map((k) => [k, v[k] as Json] as const));
}

export function nodeFromJson(v: Json): Node {
  if (!isJsonObject(v)) throw new NodeDecodeError("expected a JSON object for a node");
  const type = reqString("node", "type", v);
  switch (type) {
    case "text":
      return { t: "text", value: req("text node", "value", v), annotations: reqAnnotations(v) };
    case "element":
      return {
        t: "element",
        tag: reqString("element node", "tag", v),
        attributes: reqArray("element node", "attributes", v).map(nodeAttributeFromJson),
        value: req("element node", "value", v),
        children: reqArray("element node", "children", v).map(nodeFromJson),
        annotations: reqAnnotations(v),
      };
    case "fragment":
      return {
        t: "fragment",
        children: reqArray("fragment node", "children", v).map(nodeFromJson),
        annotations: reqAnnotations(v),
      };
    default:
      throw new NodeDecodeError(`unknown node type: ${JSON.stringify(type)}`);
  }
}

export function nodeAttributeFromJson(v: Json): NodeAttribute {
  if (!isJsonObject(v)) throw new NodeDecodeError("expected a JSON object for a node attribute");
  const kind = reqString("node attribute", "kind", v);
  switch (kind) {
    case "attribute":
      return {
        t: "attribute",
        name: reqString("attribute", "name", v),
        value: req("attribute", "value", v),
      };
    case "action":
      return {
        t: "action",
        event: reqString("action", "event", v),
        key: reqString("action", "key", v),
        payload: req("action", "payload", v),
      };
    default:
      throw new NodeDecodeError(`unknown node attribute kind: ${JSON.stringify(kind)}`);
  }
}

/** Rewrites every action reachable in a tree, leaving everything else untouched. */
export function mapActions(
  n: Node,
  f: (event: string, key: string, payload: Json) => NodeAttribute,
): Node {
  switch (n.t) {
    case "text":
      return n;
    case "element":
      return {
        ...n,
        attributes: n.attributes.map((a) =>
          a.t === "action" ? f(a.event, a.key, a.payload) : a,
        ),
        children: n.children.map((c) => mapActions(c, f)),
      };
    case "fragment":
      return { ...n, children: n.children.map((c) => mapActions(c, f)) };
  }
}

File: tramaj-py/tramaj/node.py
"""``Node``/``NodeAttribute`` and the strict ``specs/node-json.md`` encode/decode,
hand-written so the "a decoder MUST reject ... MUST NOT infer a missing field
from a default" rules are enforced exactly."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Callable, Union

from .jsonval import Json

Annotations = dict


def no_annotations() -> Annotations:
    return {}


@dataclass
class TextNode:
    value: Json
    annotations: Annotations = field(default_factory=dict)
    t: str = field(default="text", init=False)


@dataclass
class ElementNode:
    tag: str
    attributes: list["NodeAttribute"]
    value: Json
    children: list["Node"]
    annotations: Annotations = field(default_factory=dict)
    t: str = field(default="element", init=False)


@dataclass
class FragmentNode:
    children: list["Node"]
    annotations: Annotations = field(default_factory=dict)
    t: str = field(default="fragment", init=False)


Node = Union[TextNode, ElementNode, FragmentNode]


@dataclass
class AttributeAttr:
    name: str
    value: Json
    t: str = field(default="attribute", init=False)


@dataclass
class ActionAttr:
    event: str
    key: str
    payload: Json
    t: str = field(default="action", init=False)


NodeAttribute = Union[AttributeAttr, ActionAttr]


def text_node(value: Json, annotations: Annotations | None = None) -> Node:
    return TextNode(value, {} if annotations is None else annotations)


def element_node(
    tag: str,
    attributes: list[NodeAttribute],
    value: Json,
    children: list[Node],
    annotations: Annotations | None = None,
) -> Node:
    return ElementNode(tag, attributes, value, children, {} if annotations is None else annotations)


def fragment_node(children: list[Node], annotations: Annotations | None = None) -> Node:
    return FragmentNode(children, {} if annotations is None else annotations)


def node_to_json(n: Node) -> Json:
    if n.t == "text":
        return {"type": "text", "value": n.value, "annotations": dict(n.annotations)}
    if n.t == "element":
        return {
            "type": "element",
            "tag": n.tag,
            "attributes": [node_attribute_to_json(a) for a in n.attributes],
            "value": n.value,
            "children": [node_to_json(c) for c in n.children],
            "annotations": dict(n.annotations),
        }
    return {
        "type": "fragment",
        "children": [node_to_json(c) for c in n.children],
        "annotations": dict(n.annotations),
    }


def node_attribute_to_json(a: NodeAttribute) -> Json:
    if a.t == "attribute":
        return {"kind": "attribute", "name": a.name, "value": a.value}
    return {"kind": "action", "event": a.event, "key": a.key, "payload": a.payload}


class NodeDecodeError(Exception):
    pass


def _req(what: str, fieldname: str, obj: dict) -> Json:
    if fieldname not in obj:
        raise NodeDecodeError(f'{what}: missing required field "{fieldname}"')
    return obj[fieldname]


def _req_string(what: str, fieldname: str, obj: dict) -> str:
    v = _req(what, fieldname, obj)
    if not isinstance(v, str):
        raise NodeDecodeError(f'{what}: field "{fieldname}" must be a string')
    return v


def _req_array(what: str, fieldname: str, obj: dict) -> list:
    v = _req(what, fieldname, obj)
    if not isinstance(v, list):
        raise NodeDecodeError(f'{what}: field "{fieldname}" must be an array')
    return v


def _req_annotations(obj: dict) -> Annotations:
    v = _req("node", "annotations", obj)
    if not isinstance(v, dict):
        raise NodeDecodeError('node: field "annotations" must be an object')
    return dict(v)


def node_from_json(v: Json) -> Node:
    if not isinstance(v, dict):
        raise NodeDecodeError("expected a JSON object for a node")
    typ = _req_string("node", "type", v)
    if typ == "text":
        return TextNode(_req("text node", "value", v), _req_annotations(v))
    if typ == "element":
        return ElementNode(
            _req_string("element node", "tag", v),
            [node_attribute_from_json(a) for a in _req_array("element node", "attributes", v)],
            _req("element node", "value", v),
            [node_from_json(c) for c in _req_array("element node", "children", v)],
            _req_annotations(v),
        )
    if typ == "fragment":
        return FragmentNode(
            [node_from_json(c) for c in _req_array("fragment node", "children", v)],
            _req_annotations(v),
        )
    raise NodeDecodeError(f'unknown node type: "{typ}"')


def node_attribute_from_json(v: Json) -> NodeAttribute:
    if not isinstance(v, dict):
        raise NodeDecodeError("expected a JSON object for a node attribute")
    kind = _req_string("node attribute", "kind", v)
    if kind == "attribute":
        return AttributeAttr(_req_string("attribute", "name", v), _req("attribute", "value", v))
    if kind == "action":
        return ActionAttr(
            _req_string("action", "event", v),
            _req_string("action", "key", v),
            _req("action", "payload", v),
        )
    raise NodeDecodeError(f'unknown node attribute kind: "{kind}"')


def map_actions(n: Node, f: Callable[[str, str, Json], NodeAttribute]) -> Node:
    """Rewrites every action reachable in a tree, leaving everything else untouched."""
    if n.t == "text":
        return n
    if n.t == "element":
        return ElementNode(
            n.tag,
            [f(a.event, a.key, a.payload) if a.t == "action" else a for a in n.attributes],
            n.value,
            [map_actions(c, f) for c in n.children],
            n.annotations,
        )
    return FragmentNode([map_actions(c, f) for c in n.children], n.annotations)
